geode-redis/src/main/java/org/apache/geode/redis/internal/executor/string/SetExecutor.java

```diff
@@ -35,105 +37,182 @@
       return RedisResponse.string(SUCCESS);
+    } else {
+      return RedisResponse.nil();
     }
-
-    return RedisResponse.nil();
   }
 
   private ByteArrayWrapper getValueToSet(List<byte[]> commandElems) {
     byte[] value = commandElems.get(2);
     return new ByteArrayWrapper(value);
   }
 
+  private SetOptions parseOptionalParameters(List<byte[]> optionalParameterBytes)
+      throws IllegalArgumentException {
 
-  private SetOptions parseCommandElems(List<byte[]> commandElems) throws IllegalArgumentException {
     boolean keepTTL = false;
     SetOptions.Exists existsOption = SetOptions.Exists.NONE;
     long expiration = 0L;
 
-    for (int i = 3; i < commandElems.size(); i++) {
-      String current_arg = Coder.bytesToString(commandElems.get(i)).toUpperCase();
-      switch (current_arg) {
-        case "KEEPTTL":
-          keepTTL = true;
-          break;
-        case "EX":
-          if (expiration != 0) {
-            throw new IllegalArgumentException(ERROR_SYNTAX);
-          }
-          i++;
-          expiration = parseExpirationTime(i, commandElems);
-          expiration = SECONDS.toMillis(expiration);
-          break;
-        case "PX":
-          if (expiration != 0) {
-            throw new IllegalArgumentException(ERROR_SYNTAX);
-          }
-          i++;
-          expiration = parseExpirationTime(i, commandElems);
-          break;
-        case "NX":
-          if (existsOption != SetOptions.Exists.NONE) {
-            throw new IllegalArgumentException(ERROR_SYNTAX);
-          }
-          existsOption = SetOptions.Exists.NX;
-          break;
-        case "XX":
-          if (existsOption != SetOptions.Exists.NONE) {
-            throw new IllegalArgumentException(ERROR_SYNTAX);
-          }
-          existsOption = SetOptions.Exists.XX;
-          break;
-        default:
-          throw new IllegalArgumentException(ERROR_SYNTAX);
+    List<String> optionalParametersStrings =
+        optionalParameterBytes.stream()
+            .map(item -> Coder.bytesToString(item).toUpperCase())
+            .collect(Collectors.toList());
+
+    throwExceptionIfUnknownParameter(optionalParametersStrings);
+    throwExceptionIfIncompatableParamaterOptions(optionalParametersStrings);
+    keepTTL = optionalParametersStrings.contains("KEEPTL");
+
+    if (optionalParametersStrings.contains("PX")) {
+      String nextParameter =
+          getNextParameter("PX", optionalParametersStrings);
+
+      if (!isANumber(nextParameter)) {
+        throw new IllegalArgumentException(ERROR_SYNTAX);
+      }
+
+      expiration = parseExpirationTime("PX", nextParameter);
+      if (expiration <= 0) {
+        throw new IllegalArgumentException(ERROR_INVALID_EXPIRE_TIME);
+      }
+
+    } else if (optionalParametersStrings.contains("EX")) {
+      String nextParameter =
+          getNextParameter("EX", optionalParametersStrings);
+      if (!isANumber(nextParameter)) {
+        throw new IllegalArgumentException(ERROR_SYNTAX);
```

Comment:
The old code used ERROR_NOT_INTEGER if it was not a number. You have changed that to ERROR_SYNTAX

Reply:
you're right-  just checked native redis.  I'll look into lack of test here, too